xpti/include/xpti/xpti_trace_framework.hpp

```diff
@@ -479,84 +559,930 @@ inline std::string readMetadata(const metadata_t::value_type &MD) {
+    if (MData) {
+      // Set the tracepoint scope with the prepared data so all nested functions
+      // will have access to it; this call also sets the Universal ID separately
+      // for backward compatibility.
+      xptiSetTracepointScopeData(MData);
+      // Set the trace event for this tracepoint; all notifications will use
+      // this
+      MTraceEvent = MData->event_ref();
+      // Set the top flag to true, indicating that this is the top-level
+      // tracepoint.
+      MTop = true;
+    }
+  }
+
+  /// @brief Destructor for the tracepoint_scope_t class.
+  ///
+  /// @details This function is responsible for cleaning up when an instance of
+  /// tracepoint_scope_t is no longer needed. It notifies subscribers if scoped
+  /// notification is enabled and the trace type is set. If self notification
+  /// flag is set, it notifies the the end of the tracepoint scope. If this
+  /// instance is the top-level tracepoint, it releases the event, unsets the
+  /// tracepoint scope data, resets the data member, and sets the top flag to
+  /// false.
+  ///
+  ~tracepoint_scope_t() {
+    // If scoped notification is enabled and the trace type is set, notify
+    // subscribers with the closing trace type by setting the LSB to 1 so it
+    // will correspond to the the one specified in scopedNotify(). We will check
+    // if the trace type has a scope defined before notifying.
+    if (MScopedNotify && MTraceType && isTraceTypeScoped()) {
+
+      MTraceType = MTraceType | 1;
+      if (xptiCheckTraceEnabled(MStreamId, MTraceType))
+        xptiNotifySubscribers(MStreamId, MTraceType, MParentEvent, MTraceEvent,
+                              MScopedCorrelationId, MScopedUserData);
+    }
+    // If self notify flag is set, notify the end of the scope
+    selfNotifyEnd();
+    // If this is the top-level tracepoint, perform additional cleanup.
+    if (MTop) {
+      // Release the tracepoint created since this instance of the payload is
+      // going out of scope
+      xptiDeleteTracepoint(MData);
+      // Reset TLS data to invalid
+      xptiUnsetTracepointScopeData();
+      // Clear internal state
+      MData = nullptr;
+      MTop = false;
+    }
+  }
+
+  /// @brief Returns the 64-bit unique identifier (UID) of the tracepoint.
+  ///
+  /// This function returns a reference to the UID of the tracepoint, which is
+  /// stored in the tracepoint's data. This is created when the tracepoint is
+  /// created using the payload or inherited through the TLS
+  ///
+  /// @return A reference to the UID of the tracepoint.
+  ///
+  uint64_t uid64() { return MData->uid64(); }
+
+  /// @brief Returns the payload of the tracepoint.
+  ///
+  /// This function returns a pointer to the payload of the tracepoint, which is
+  /// which corresponds to the UID in the trace point data.
+  ///
+  /// @return A pointer to the payload of the tracepoint.
+  ///
+  xpti::payload_t *payload() { return MData->payload_ref(); }
```

Comment:
just a nit: I am confused having method "payload_ref" that returns ptr actually

Reply:
Was forward looking to eventually move to shared_ptr() - prefer not to change as this is the ABI we want to settle on moving forward.